### ISSUE TITLE
A better way to represent iowait

### ISSUE THREAD

**Issue by UltimateByte** [netdata/netdata#2501]
Hi everywone. While setting up and testing a NAS recently, I came up with weird feelings about the way iowait is represented in Netdata. So I did some research about it when it occurred, and for some reason, this morning, I just magically woke up with some ideas about it.

### What is iowait
Iowait is the time a CPU process is waiting for HDD in order to complete a task. As a matter of fact, iowait is idle time, with the subtle nuance that more CPU could be used by the process causing iowait, if iowait wasn't occurring.

### Suggestions
iowait should be considered as idle time, it has to remain very visible since it is very important, but it shouldn't be misleading in the way that it makes you think your CPU is unusable because there is iowait, which it's not. Because something is HDD limited doesn't mean the whole system is HDD limited, especially if you got multiple HDDs.
For that matter:
- iowait should be represented with grey/transparent or dotted line onto the CPU graph
- iowait shouldn't be counted in the total CPU usage, or if it is, it should be on a secondary layer with the same different color/drawing than onto the main CPU graph.

### For skeptical people: Proof of iowait not being real CPU usage
- Make an HDD intensive task (like file copy).
- Monitor read/write speed using netdata, and monitor iowait and Netdata CPU usage that should increase. 
- Now in the same time, do some CPU intensive tasks (like sysbench): iowait should decrease and the copy rate shouldn't be affected much if at all, while CPU usage should increase or remain about the same, depending of the threads you set sysbench to run at and the original iowait that you had.
- Watch your sysbench result, then stop the copy, and redo a sysbench test: you should get around the same performance resut, because sysbench wasn't affected by iowait.

Thank you.

**Comment by ktsaou** [comment]
Hi, of course it is like that.

Keep in mind though that the total CPU time of the system is also affected by more factors. For example, since all processors run at multiple clock rates, the cpu time % reported can be affected by these clocks too. 100% at 2GHz is not the same with 100% at 4GHz and the system may be capable of reaching 4GHz.

You will experience this if you monitor the CPU time reported at the Applications or Systemd Services charts. You may have an application consuming 5% CPU time when the system is idle and 1% when the system is very busy. It drops because when the system is busy the clock rate of CPU increases and the application in question now runs a lot faster.

Today, at the top gauge CPU chart of the dashboard we cannot add more detail (gauge.js can do it, but it needs more work from us). Until we do it, it is not right to remove waitio from it. Removing it, will give totally wrong information. The kernel folks added waitio into the CPU to provide an indication of something  important happening. So, I think that until we enhance that gauge to visualise more metrics at one, we should keep it like it is now.

**Comment by BrainDoctor** [comment]
I'm not so sure about this. netdata's iowait dimension actually helped me tracking down a performance issue on two of my servers. Every minute, the systems were at 100% CPU usage for about 5 to 6 seconds. they could still complete other tasks, but it was definitely noticable when using disks.

Here's why:
Both servers have attached a modern USB3 drive each. However, ext4 is installed and by default journaling is enabled. I somehow found out that the journaling process for the usb drive was stalling and other disks too (those with ext4, not ZFS ones). Because of driver and motherboard problems I had to put them in USB2 ports. They aren't used much and are totally okay for the workload. 

The problem:
with the command I found somewhere on Stackoverflow `while true; do date; ps auxf | awk '{if($8=="D") print $0;}'; sleep 1; done` I found out that the **journaling** process was waiting or hanging during these 5 seconds. I believe the "D" flag means that it's an uninterruptible process (not so sure, I'm no expert). The drive is fine, but for some reason I don't understand, the journaling process was stalling the system and additionally blocking any disk access for that drive. Could it be that the journaling process is not directly dependent on the disk, but stills shows up in iowait?

The solution:
I spent nearly 2 days to figure it out and had to **completely disable ext4's Journaling** and got about 15% reduction in cpu usage (Intel Xeon E3-1225 v3 @ 3.20GHz). Also the peaks every minute is now negligible and caused by some kworker threads (which is fine).

TL;DR
I'm totally happy the way it is represented now. If a system is behaving strange, I want to know why. If netdata is hiding such problems for UI reasons, then it failed its core task. If iowait is displayed differently I'm okay with it, but not if its hidden or confusing. A system is only so fast as its slowest member :)

**Comment by UltimateByte** [comment]
@ktsaou Yes, % CPU usage depends on CPU throttling, but it's not comparable to idle or iowait, even if i see why you see a relation.

If the top gauche is hard to manage for now, then at least it would be cool to put the iowait right under softirq or before softirq, and to make it white or light grey.

As for the top gauche, what i would love to see is one big neddle for the main CPU usage, and another one that would be smaller and whiter for "other cpu time" including softirq and iowait. That way, the info is still there and there is no confusion possible.

Also, the statement "Total CPU utilization (all cores). 100% here means there is no CPU idle time at all." isn't right if we agree that iowait is idle time and since iowait is counted into total CPU usage.

**Comment by Ferroin** [comment]
This really boils down to whether or not you consider waiting on something to be a busy state.  From a processing perspective, it can be argued that it is a busy state because the application that triggered the iowait state (seen as D state in top and related tools) can't do anything else until the IO completes.  If that's your only thing running, then your system _is_ busy, the CPU just isn't doing anything except servicing interrupts.

IMO, this really falls into the same category as load average, it's useful in certain specific circumstances, but not much outside of that.  The statement on the netdata dashboard about iowait is accurate, but high iowait could also mean that whatever you're running isn't efficiently parallelizing (or that it can't do so).

**Comment by paulfantom** [comment]
Closing this due to prolonged inactivity. Please reopen if you think this is still relevant.

**Comment by UltimateByte** [comment]
After a couple of years, I find it convenient to have I/O showing some CPU usage, it usually correlates to load and helps seing system bottlenecks.

So you were all right, nothing to change here. 
Thank you all for taking time to discuss this suggestion.
